File: client/dto/audit.go
package dto

var AllowedActions = map[string]struct{}{
	"CREATE":                  {},
	"UPDATE":                  {},
	"RESTORE":                 {},
	"DELETE":                  {},
	"FORCE_DELETE":            {},
	"UPSERT":                  {},
	"INVITE":                  {},
	"RESEND_INVITE":          {},
	"REVOKE_INVITE":          {},
	"ADD_COLLABORATOR":       {},
	"REMOVE_COLLABORATOR":    {},
	"CREATE_TOKEN":           {},
	"REVOKE_TOKEN":           {},
	"LOGIN":                  {},
	"LOGIN2FA":               {},
	"UNSUCCESSFUL_LOGIN":     {},
	"ADD_MEMBERSHIP":         {},
	"REMOVE_MEMBERSHIP":      {},
	"ERROR_BUDGET_RESET":     {},
	"START":                  {},
	"END":                    {},
	"STAGE_START":            {},
	"STAGE_END":              {},
	"PAUSE":                  {},
	"RESUME":                 {},
	"ABORT":                  {},
	"TIMEOUT":                {},
	"SIGNED_EULA":            {},
	"ROLE_ASSIGNMENT_CREATED": {},
	"ROLE_ASSIGNMENT_UPDATED": {},
	"ROLE_ASSIGNMENT_DELETED": {},
	"MOVE":                   {},
	"ENABLED":                {},
	"DISABLED":               {},
	"DISMISS_ANOMALY":        {},
	"RERUN":                  {},
	"BYPASS":                 {},
	"STABLE_VERSION_CHANGED": {},
	"SYNC_START":             {},
	"START_IMPERSONATION":    {},
	"END_IMPERSONATION":      {},
	"MOVE_TO_GIT":            {},
	"FREEZE_BYPASS":          {},
	"EXPIRED":                {},
	"FORCE_PUSH":             {},
}

type AuditPrincipal struct {
	Type       string `json:"type"`
	Identifier string `json:"identifier"`
}

type AuditResourceScope struct {
	AccountIdentifier string `json:"accountIdentifier"`
	OrgIdentifier     string `json:"orgIdentifier,omitempty"`
	ProjectIdentifier string `json:"projectIdentifier,omitempty"`
}

type ListAuditEventsFilter struct {
	Scopes     []AuditResourceScope `json:"scopes,omitempty"`
	Principals []AuditPrincipal     `json:"principals,omitempty"`
	Actions    []string             `json:"actions,omitempty"`
	FilterType string               `json:"filterType,omitempty"`
	StartTime  int64                `json:"startTime,omitempty"`
	EndTime    int64                `json:"endTime,omitempty"`
}

type AuditOutput[T any] struct {
	Status string             `json:"status,omitempty"`
	Data   AuditOutputData[T] `json:"data,omitempty"`
}

type AuditOutputData[T any] struct {
	PageItemCount int    `json:"pageItemCount,omitempty"`
	PageSize      int    `json:"pageSize,omitempty"`
	Content       []T    `json:"content,omitempty"`
	PageIndex     int    `json:"pageIndex,omitempty"`
	HasNext       bool   `json:"hasNext,omitempty"`
	PageToken     string `json:"pageToken,omitempty"`
	TotalItems    int    `json:"totalItems,omitempty"`
	TotalPages    int    `json:"totalPages,omitempty"`
}

type AuditListItem struct {
	AuditID            string                  `json:"auditId,omitempty"`
	InsertId           string                  `json:"insertId,omitempty"`
	Resource           AuditResource           `json:"resource,omitempty"`
	Action             string                  `json:"action,omitempty"`
	Module             string                  `json:"module,omitempty"`
	Timestamp          int64                   `json:"timestamp,omitempty"`
	AuthenticationInfo AuditAuthenticationInfo `json:"authenticationInfo,omitempty"`
	ResourceScope      AuditResourceScope      `json:"resourceScope,omitempty"`
	Time               string
}

type AuditResource struct {
	Type       string              `json:"type"`
	Identifier string              `json:"identifier"`
	Labels     AuditResourceLabels `json:"labels,omitempty"`
}

type AuditResourceLabels struct {
	ResourceName string `json:"resourceName,omitempty"`
}

type AuditAuthenticationInfo struct {
	Principal AuditPrincipal                `json:"principal"`
	Labels    AuditAuthenticationInfoLabels `json:"labels,omitempty"`
}

type AuditAuthenticationInfoLabels struct {
	UserID   string `json:"userId,omitempty"`
	Username string `json:"username,omitempty"`
}

File: src/registry/toolsets/audit.ts
import type { ToolsetDefinition } from "../types.js";
import { ngExtract, pageExtract } from "../extractors.js";

/** Parse ISO 8601 to Unix ms. Returns NaN if invalid. */
function parseIsoToMs(value: unknown): number {
  if (value === undefined || value === "") return NaN;
  const t = new Date(String(value)).getTime();
  return Number.isNaN(t) ? NaN : t;
}

/** Default 7-day window (matches v1 list_user_audits). */
function defaultAuditTimeWindow(): { startTime: number; endTime: number } {
  const endTime = Date.now();
  const startTime = endTime - 7 * 24 * 60 * 60 * 1000;
  return { startTime, endTime };
}

export const auditToolset: ToolsetDefinition = {
  name: "audit",
  displayName: "Audit Trail",
  description: "Platform audit events and activity log",
  resources: [
    {
      resourceType: "audit_event",
      displayName: "Audit Event",
      description: "Audit trail event. Supports list and get (YAML diff).",
      toolset: "audit",
      scope: "account",
      identifierFields: ["audit_id"],
      listFilterFields: [
        { name: "resource_type", description: "Filter audit logs by resource type", enum: ["ORGANIZATION", "PROJECT", "USER_GROUP", "SECRET", "PIPELINE", "TRIGGER", "TEMPLATE", "INPUT_SET", "DELEGATE_CONFIGURATION", "DELEGATE_GROUPS", "SERVICE", "ENVIRONMENT", "ENVIRONMENT_GROUP", "DELEGATE", "SERVICE_ACCOUNT", "CONNECTOR", "ROLE", "RESOURCE_GROUP", "DASHBOARD", "GOVERNANCE_POLICY", "GOVERNANCE_POLICY_SET", "VARIABLE", "MONITORED_SERVICE", "FEATURE_FLAG", "CHAOS_HUB", "CHAOS_INFRASTRUCTURE", "CHAOS_EXPERIMENT", "GITOPS_AGENT", "GITOPS_APPLICATION", "CODE_REPOSITORY", "SETTING", "DEPLOYMENT_FREEZE"] },
        { name: "action", description: "Filter audit logs by action type", enum: ["CREATE", "UPDATE", "RESTORE", "DELETE", "FORCE_DELETE", "UPSERT", "INVITE", "RESEND_INVITE", "REVOKE_INVITE", "ADD_COLLABORATOR", "REMOVE_COLLABORATOR", "CREATE_TOKEN", "REVOKE_TOKEN", "LOGIN", "LOGIN2FA", "UNSUCCESSFUL_LOGIN", "ADD_MEMBERSHIP", "REMOVE_MEMBERSHIP", "START", "END", "PAUSE", "RESUME", "ABORT", "TIMEOUT", "ROLE_ASSIGNMENT_CREATED", "ROLE_ASSIGNMENT_UPDATED", "ROLE_ASSIGNMENT_DELETED", "ENABLED", "DISABLED", "RERUN", "BYPASS"] },
        { name: "start_time", description: "Start time in ISO 8601 format (e.g. 2025-07-10T08:00:00Z). Default: 7 days ago." },
        { name: "end_time", description: "End time in ISO 8601 format (e.g. 2025-07-10T23:59:59Z). Default: now." },
        { name: "search_term", description: "Filter audit logs by search term" },
        { name: "module", description: "Filter audit logs by module" },
      ],
      deepLinkTemplate: "/ng/account/{accountId}/settings/audit-trail",
      operations: {
        list: {
          method: "POST",
          path: "/audit/api/audits/list",
          queryParams: { page: "pageIndex", size: "pageSize" },
          bodyBuilder: (input) => {
            const { startTime: defaultStart, endTime: defaultEnd } = defaultAuditTimeWindow();
            const startMs = parseIsoToMs(input.start_time);
            const endMs = parseIsoToMs(input.end_time);
            const startTime = Number.isNaN(startMs) ? defaultStart : startMs;
            const endTime = Number.isNaN(endMs) ? defaultEnd : endMs;
            return {
              filterType: "Audit",
              modules: input.module ? [input.module] : undefined,
              actions: input.action ? [input.action] : undefined,
              startTime,
              endTime,
            };
          },
          responseExtractor: pageExtract,
          description: "List audit events",
        },
        get: {
          method: "GET",
          path: "/audit/api/auditYaml",
          queryParams: { audit_id: "auditId" },
          responseExtractor: ngExtract,
          description: "Get audit event YAML diff",
        },
      },
    },
  ],
};
